src/redux/middlewares/CentrifugoMiddleware/index.js

```diff
@@ -16,81 +15,137 @@ import {
+  }
+
+  return action.payload === 'active';
+}
+
+function shouldDisconnectBasedOnAppState(getState, action) {
+  if (!shouldManageConnectionBasedOnAppState(getState, action)) {
+    return false;
+  }
+
+  if (action.type !== appStateChanged.type) {
+    return false;
+  }
+
+  return action.payload !== 'active';
+}
+
+let centrifuge = null;
+let subscription = null;
 
+export default ({ getState, dispatch }) => {
   return next => action => {
     if (action.type === LOGOUT_SUCCESS) {
-      const result = next(action);
+      dispatch(disconnect());
+      return next(action);
+    }
+
+    if (
+      action.type === connect.type ||
+      shouldConnectBasedOnAppState(getState, action)
+    ) {
+      const state = getState();
+
+      if (
+        !selectIsAuthenticated(state) ||
+        !selectHttpClientHasCredentials(state)
+      ) {
+        return next(action);
+      }
 
       if (centrifuge && centrifuge.isConnected()) {
-        centrifuge.disconnect();
-        centrifuge = null;
+        return next(action);
       }
 
-      return result;
-    }
+      const httpClient = selectHttpClient(state);
+      const baseURL = selectBaseURL(state);
+      const user = selectUser(state);
+
+      httpClient.get('/api/centrifugo/token').then(tokenResponse => {
+        const url = parseUrl(baseURL);
+        const protocol = url.protocol === 'https:' ? 'wss' : 'ws';
+
+        centrifuge = new Centrifuge(
+          `${protocol}://${url.hostname}/centrifugo/connection/websocket`,
+          {
+            debug: __DEV__,
+            onRefresh: function (ctx, cb) {
+              httpClient
+                .post('/api/centrifugo/token/refresh')
+                .then(refreshResponse => {
+                  // @see https://github.com/centrifugal/centrifuge-js#refreshendpoint
+                  // Data must be like {"status": 200, "data": {"token": "JWT"}} - see
+                  // type definitions in dist folder. Note that setting status to 200 is
+                  // required at moment. Any other status will result in refresh process
+                  // failure so client will eventually be disconnected by server.
+                  cb({ status: 200, data: { token: refreshResponse.token } });
+                });
+            },
+          },
+        );
+
+        centrifuge.setToken(tokenResponse.token);
+
+        centrifuge.on('connect', context => dispatch(connected(context)));
+        centrifuge.on('disconnect', context => dispatch(disconnected(context)));
+
+        subscription = centrifuge.subscribe(
+          `${tokenResponse.namespace}_events#${user.username}`,
+          msg => dispatch(message(msg.data.event)),
+        );
+
+        centrifuge.connect();
+      });
 
-    if (!isCentrifugoAction(action)) {
       return next(action);
     }
 
-    const state = getState();
-
     if (
-      !selectIsAuthenticated(state) ||
-      !selectHttpClientHasCredentials(state)
+      action.type === disconnect.type ||
```

Comment:
shouldnt it be `&&` ?

Reply:
No, that's correct. We want to proceed only if both are true. To be honest, I think the second check is redundant, but I don't want to experiment with it now 🙈

Reply:
that's why i am wondering it should be AND here? instead of OR

Reply:
The check is if NOT (`selectIsAuthenticated` AND `selectHttpClientHasCredentials`), which is the same as (NOT `selectIsAuthenticated`) OR (NOT `selectHttpClientHasCredentials`)

Reply:
you are speaking about the removed code no?